Implement `ChannelListing` for new chat design

jai-x opened the pull request:
Part of https://github.com/ppy/osu/issues/17001

Reference design: https://www.figma.com/file/f8b2dHp9LJCMOqYP4mdrPZ/Client%2FChat?node-id=2%3A63

https://user-images.githubusercontent.com/8362491/157338056-cfbb2079-04b0-4610-a335-17fceaee889a.mp4

Adds components `ChannelListing` and `ChannelListing` item with visual
test. Essentially a more simplified version of the existing
`ChannelSelectionOverlay` component.

Correctly implements `IFilterable` behaviour to filter child channel
items. Channel joined state is based on the underlying `Joined` bindable
of the `Channel` class.

Channel join/leave events are exposed via `OnRequestJoin` and
`OnRequestLeave` events which should be handled by parent component.

Requires a cached `OverlayColourScheme` instance to be provided by the
parent overlay component when added. Uses colour scheme `Pink` in
visual tests.

For now I have put the new components under the `NewChat` namespace
and plan to remove that namespace when properly integrating into the main
game. If this is not desired I can remove it.

jai-x committed:
Implement `ChannelListing` for new chat design

Adds components `ChannelListing` and `ChannelListing` item with visual
test. Essentially a more simplified version of the existing
`ChannelSelectionOverlay` component.

Correctly implements `IFilterable` behaviour to filter child channel
items. Channel joined state is based on the underlying `Joined` bindable
of the `Channel` class.

Channel join/leave events are exposed via `OnRequestJoin` and
`OnRequestLeave` events which should be handled by parent component.

Requires a cached `OverlayColourScheme` instance to be provided by the
parent overlay component when added.

bdach commented:
Not sure about the namespacing choices here with `Online.NewChat` - generally there's no precedent for this, and what's the plan for old pieces? Is `NewChat` going to be there until the end of time?

Looks a bit funky in test browser, too:

![2022-03-08-234630_248x69_scrot](https://user-images.githubusercontent.com/20418176/157338775-7d5e5f28-361e-482e-b884-0dad577d13c3.png)

bdach commented on osu.Game/Overlays/NewChat/ChannelListing.cs:
```diff
@@ -0,0 +1,85 @@
+            }
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
```
BDL should be where the ctor is right now

jai-x commented:
For now I have put the new components under the `NewChat` namespace and plan to remove that namespace when all the components have been added the overlay is properly integrated into the main game. I mainly did this just to make things clearer for me personally while developing. If this is not desired I can remove it.

bdach commented on osu.Game/Overlays/NewChat/ChannelListingItem.cs:
```diff
@@ -0,0 +1,175 @@
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
```
order all over the place here too - preferred ordering is:

* ctor
* BDL
* `LoadComplete()`
* event handlers/protected methods
* private methods

Also see: https://github.com/ppy/osu-framework/issues/2278

jai-x replied:
I didn't know about this ordering guide. I will make sure to use it.

bdach reviewed (changes requested):
aside from code quality inspections

jai-x committed:
Reorder class attributes

jai-x committed:
Use correct fonts and colours in ChannelListingItem

jai-x commented on osu.Game/Overlays/NewChat/ChannelListingItem.cs:
```diff
@@ -0,0 +1,175 @@
+                                Text = channel.Topic,
+                                Font = OsuFont.Torus.With(size: TEXT_SIZE),
+                                Margin = new MarginPadding { Bottom = 2 },
+                                Colour = Colour4.White,
```
`Content1` seemed to resolve to white `RGB(1,1,1)` so I've used `Content2` instead.

jai-x commented on osu.Game/Overlays/NewChat/ChannelListing.cs:
```diff
@@ -0,0 +1,85 @@
+                        Spacing = new Vector2(3),
+                        Padding = new MarginPadding
+                        {
+                            Vertical = 13,
+                            Horizontal = 15,
```
Would it be better to adjust sizing now or keep the sizing verbatim for now and make a post-pass after all the components have been added?

bdach replied:
I personally do it as I open PRs for the individual pieces. A giant downsizing PR at the end of everything may not be super great, I dunno.

jai-x committed:
Use `int` in ChannelListing test

peppy commented:
On the topic of namespace, do you actually have any conflicting class names? In this PR it looks like the naming you've chosen differs from existing classes, at which point I'd not create a new namespace at all if that's feasible.

Rationale is that following git changes across files is a bit of a pain so where possible we should try and put files in their final home from the outset.

jai-x committed:
Remove `NewChat` namespace

peppy committed:
Merge branch 'master' into new-chat-channel-listing

peppy committed:
Add animation to checkbox when joning/leaving a channel

peppy committed:
Move `ChannelListingItem` spacing into item so input is always handled by an item in the list

Without this change, there would be a couple of pixels between each list
item where nothing would be hovered. This is a pretty annoying UX which
we should be avoiding we possible.

peppy commented:
I've made a couple of minor tweaks, explained in the commit messages.

Looks good otherwise!
